### PR TITLE
Handle endpoints with the same name

### PR DESCRIPTION
**Pull request by rjschwei**
AWS has introduced duplication in endpoint names in the metadata such as iam/info and identity-credentials/ec2/info. Previously the first endpoint found would get added to the CLI options and the second one would get dropped. The new code will use the path to differentiate the query options and collect all the data when no query option is specified.

### CODE REVIEW

**Review by bear454 — APPROVED** [review]
Thanks for the quick fix.

File: lib/ec2metadata/__init__.py [hunk 1]
```diff
@@ -68,11 +68,28 @@ def _add_mata_option(self, path):
             if item:
                 if item == 'public-keys/':
                     continue
-                if item not in options:
-                    if item[-1] != '/':
+                if item[-1] == '/':
+                    self._add_mata_option(path+item)
```
Comment (smarlowucf) [inline lib/ec2metadata/__init__.py]: Pre-existing typo: s/mata/meta

File: lib/ec2metadata/__init__.py [hunk 2]
```diff
@@ -68,11 +68,28 @@ def _add_mata_option(self, path):
...
+                        # Construct a new name for the option using the given
+                        # path as name addition
+                        option_name = item
+                        # Drop the empty entry, the path has a '/' at the end
+                        prefix_name_opts = path.split('/')[:-1]
+                        prefix_name_opts.reverse()
+                        cnt = 1
+                        while option_name in options:
+                            option_name = '-'.join(
+                                prefix_name_opts[:cnt] + [option_name]
```
Comment (smarlowucf) [inline lib/ec2metadata/__init__.py]: Not sure I follow logic without context but it seems (could be wrong) like we end up with two options:

- info
- ec2-identity-credentials-info

If there's any clash I would expect both names become verbose and I wonder if the dash in identitycredentials causes any trouble?

Expectation would be:

- iam-info
- ec2-identity-credentials-info

Do we split this back on dash? Then identity-credentials/ec2/info becomes identity/credentials/ec2/info?

Comment (rjschwei) [reply lib/ec2metadata/__init__.py]: Well there is no way to predict if there is a clash. When the code process the first endpoint called `info` it does not know that there will be another.

If we expect the expansion of both names, which is a good suggesting that can be added but requires a bit more refactoring as we wold then have to remove the previously generated option name and insert the newly expanded name. But then if a third one comes along we are back to the short name, i.e. `info`. That can only be accounted for with an additional tracking list `duplicate_names` which in and off itself if not a problem.

But generally this is a good idea as the current approach of "first one wins the naming battle" is flaky in the sense that if IMDS changes the order in which things are presented than `--info` means one thing today and may mean a different thing tomorrow.

I'll do some refinement

**Review by smarlowucf — APPROVED** [review]

**Review by bear454 — COMMENTED** [review]
If I read correctly, `--info` will no longer be a valid argument, replaced by:

* `--iam-info`
* `--ec2-info`

Is that a correct interpretation?

**Comment by rjschwei** [comment]
> If I read correctly, `--info` will no longer be a valid argument, replaced by:
> 
> * `--iam-info`
> * `--ec2-info`
> 
> Is that a correct interpretation?

Yes that is correct. There is a pending PR for suse-rancher-setup to deal with this incompatible change.

**Pull request merged**